syntax/ffc.flex

```diff
@@ -0,0 +1,154 @@
+/* -*- Mode: bison -*-
+/*
+ *  The scanner definition for FFC.
+ */
+
+%{
+
+#include <string>
+#include <iostream>
+
+using namespace std;
+
+string string_buf;
+string operator_buf;
+
+int curr_lineno = 1;
+extern int verbose_flag;
+extern "C" int yylex();
+
+#include "tokens.hpp"
+YYSTYPE yylval;
+
+extern "C" int optind;
+
+void handle_flags(int argc, char *argv[]);
+
+FILE *fin;
+
+#undef YY_INPUT
+#define YY_INPUT(buf,result,max_size) \
+    if ((result = fread((char *)buf, 1, max_size, fin)) < 0) { \
+        YY_FATAL_ERROR("read() failed in lexer"); \
+    }
+
+extern void dump_token(int lineno, int token, YYSTYPE yylval);
+
+int main(int argc, char** argv) {
+    int token;
+
+    while (optind < argc) {
+        fin = fopen(argv[optind], "r");
+        if (fin == NULL) {
+            cerr << "Could not open input file " << argv[optind] << endl;
+            exit(1);
+        }
+
+        curr_lineno = 1;
+
+        //
+        // Scan and print all tokens.
+        //
+        cout << "#name \"" << argv[optind] << "\"" << endl;
+        while ((token = yylex()) != 0) {
+            dump_token(curr_lineno, token, yylval);
+        }
+        fclose(fin);
+        optind++;
+    }
+    exit(0);
+}
+
+%}
+%x IN_STRING
+
+INT [1-9][0-9]*
+FLOAT [0-9]+\.[0-9]+
+ID [A-z_][A-z0-9_]*
+OP [!~@#$%&^*-+\\/<>][!~@#$%&^*-+\\/<>=]*
+%%
+
+fnc        return Fnc;
+extern     return Extern;
+operator   return Operator_kw;
```

Comment:
As I understand, operator definition consists of `operator` keyword, operator itself (token if type `Operator` and value like `+` or `++` or whatever), token `(`, some optional parameters, token `)` and then the body of definition.
This together indeed constitutes `OperatorDef` AST node. But the keyword is only a small part of the definition. You don't call `fnc` 's token and `FncDef`, you call it just `fnc`.

Here we have some confusion, because we have `operator` keyword and `Operator` token, and they are distinct.

If we call the token `OperatorDef` it would be a problem, since later in bison grammar we will also have grammar rule with same name, which would include token with this name.

This would be even more confusing.

So... Do you really insist that `operator` keyword should produce token `OperatorDef` during lexical analysis?